### PR TITLE
WW-4905: Allows Initializable interface on implementing class instead of interface

### PR DESCRIPTION
**Pull request by lukaszlenart**
Implements [WW-4905](https://issues.apache.org/jira/browse/WW-4905)

### CODE REVIEW

**Comment by lukaszlenart** [comment]
@apache/struts-committers this PR fails on purpose as I think about another change - do not declare beans that are optional and can be used by users when defined in `struts.xml`

**Comment by lukaszlenart** [comment]
@apache/struts-committers ok, problem resolved - now CI will properly initialise all the beans :)

**Comment by lukaszlenart** [comment]
And how do you feel about moving this into 2.5.15? The `Initializable` interface was introduced in 2.5.14 so it would be good to fix that ASAP :)

**Comment by yasserzamani** [comment]
>  how do you feel about moving this into 2.5.15?

This is in my review queue. I'll feel when processed 😉 ( commits and changed files are somehow not few but hopefully I like to deliver this in 2.5.15 if I could dominate it )

**Comment by lukaszlenart** [comment]
one tiny issue: by "CI" I meant "DI" ;-)

**Comment by yasserzamani** [comment]
😮 but I expected failure of both `ContainerImplTest.testEarlyInitializable/testInitializable`. They correctly fail here using IntelliJ because of duplicate calls of `InitializableFactory<T>.create()`. Maybe we have to add a method `boolean isInitialized()` to your interface :) what do you think? I feel there should be a better way.

**Comment by lukaszlenart** [comment]
Not sure what you mean but there is no such tests in `ContainerImplTest` - should I add them?

**Comment by yasserzamani** [comment]
No I already pushed them at this PR's [df4a181](https://github.com/apache/struts/pull/196/commits/df4a18107f0fbf21f80cdbe8fbadc42bd2c78555#diff-81e6db967db14c2e0dd5d263e35a935f) commit :)

I discovered some tests including these are skipped in Jenkins and Travis :( with IntelliJ built-in test, I have failures here.

**Comment by lukaszlenart** [comment]
Singletons are initialized in tests early, the builder has a flag to instantiate all the singletons and this flag is  `true` in tests but `false` otherwise.

**Comment by lukaszlenart** [comment]
And see [here](https://github.com/apache/struts/blob/master/core/pom.xml#L63-L69), you can try to remove some excludes and run them locally using `mvn test`

**Comment by yasserzamani** [comment]
Thanks Lukasz :)

Please see my improved pushed tests at [ContainerImplTest.java#L114](https://github.com/lukaszlenart/struts/blob/38450c9a707c00c8c3150b5c4619d7dabe1eca59/core/src/test/java/com/opensymphony/xwork2/inject/ContainerImplTest.java#L114). I see when a singleton bean is injected twice into different objects, it's `init` method will being called twice. I think prototype scope may work but I didn't test other scopes (they may also have this issue I think).

Here are my test output:

```text
Expected :initialized
Actual   :initializedinitialized
	at com.opensymphony.xwork2.inject.ContainerImplTest.testInitializable(ContainerImplTest.java:136)
================
Expected :initialized early
Actual   :initialized earlyinitialized early
	at com.opensymphony.xwork2.inject.ContainerImplTest.testEarlyInitializable(ContainerImplTest.java:119)
```

**Comment by lukaszlenart** [comment]
Ach... I see it now ... but please ping me next time to add such tests instead of pushing them directly into my branch or at least ask upfront then I will know that I should wait for you, now I have conflicts locally :(

**Comment by yasserzamani** [comment]
Sorry for the inconvenience :( It's a good github feature that PR opener allows other changes on it. It's not only available for members but is available for all which saves time.

I'm surprised how you got conflict while I only changed a few test files which I was sure you don't change them in parallel. Do you? :)

**Comment by lukaszlenart** [comment]
I have missed your commit and modified the test in a similar way, plus fixed few other things, nvm.

**Comment by lukaszlenart** [comment]
The issue is that, Github didn't mail ma about your commits. I only noticed your comment but overlooked the commits.

**Comment by yasserzamani** [comment]
Aha... you're right. I implicitly though GitHub will email you my changes! Anyway you did great work plus near 0.1 coverage increase 👍 Thanks.

File: core/src/main/java/com/opensymphony/xwork2/inject/ContainerBuilder.java [hunk 1]
```diff
@@ -576,7 +608,9 @@ public boolean contains(Class<?> type) {
     public Container create(boolean loadSingletons) {
         ensureNotCreated();
         created = true;
-        final ContainerImpl container = new ContainerImpl(new HashMap<>(factories));
+
+        final ContainerImpl container = new ContainerImpl(factories);
```
[the remark this answers is not in the document]
Comment (yasserzamani) [reply core/src/main/java/com/opensymphony/xwork2/inject/ContainerBuilder.java]: OK, I found also another issue with prototype scope. I fixed both in one another push below and am going to merge.

**Comment by lukaszlenart** [comment]
I just see one commit 5e237e0, where is another? btw. I have got notification about it :)

**Comment by yasserzamani** [comment]
By another I meant that commit you mentioned. GitHub has changed the commit and comment order, sorry. I have no any other commit :)

**Comment by lukaszlenart** [comment]
Ok, cool :) And now I see my mistake, I had thought that you pushed into your branch not my, that's why I was surprised when the conflicts appeared.

Anyway, thanks for your help, your changes looks good 👍

**Pull request merged**